# Main entrypoint of the workflow.
# Please follow the best practices:
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there.

import pandas as pd
import yaml

# Import configuration files
params = yaml.load(open("config/params.yml", "r"), Loader=yaml.SafeLoader)
features = yaml.load(open("config/features.yml", "r"), Loader=yaml.SafeLoader)
samples = pd.read_table("config/samples.tsv", comment="#", dtype="str")


# Generate useful variables
SAMPLES = samples["sample"].drop_duplicates().values.tolist()
SAMPLE_LIB = samples[["sample", "library"]].values.tolist()
MAX_THREADS = params["max_threads"]
SNPEFF_DBS = features["snpeff"]["genome"]
CHROMOSOMES = features["reference"]["chromosomes"]


# Import subworkflows
include: "rules/folders.smk"
include: "rules/common.smk"
include: "rules/samtools.smk"
include: "rules/bcftools.smk"
include: "rules/reads.smk"
include: "rules/reference.smk"
include: "rules/fastqc.smk"
include: "rules/fastp.smk"
include: "rules/bowtie2.smk"
include: "rules/picard.smk"
include: "rules/gatk4.smk"
include: "rules/snpeff.smk"
include: "rules/report_by_sample.smk"
include: "rules/report_by_step.smk"


rule all:
    input:
        # rules.reads_link.input,
        # rules.reads_fastqc.input,
        # rules.reference.input,
        # rules.fastp.input,
        # rules.bowtie2.input,
        # rules.picard.input,
        # rules.gatk4_all.input,
        # rules.snpeff.input,
        rules.report_steps.input,
        rules.report_samples.input,
